module/os/linux/spl/spl-kmem-cache.c

```diff
@@ -160,16 +160,32 @@ taskq_t *spl_kmem_cache_taskq;		/* Task queue for aging / reclaim */
 
 static void spl_cache_shrink(spl_kmem_cache_t *skc, void *obj);
 
+typedef struct {
+	uint32_t magic;
+	int real_allocation_size;
+} kvalloc_endmarker_t;
+const uint32_t kvmagic_alive = 0xa4a4a4a4U;
+const uint32_t kvmagic_poison = 0xDEADDEADU;
+
 static void *
 kv_alloc(spl_kmem_cache_t *skc, int size, int flags)
 {
-	gfp_t lflags = kmem_flags_convert(flags);
-	void *ptr;
-
-	ptr = spl_vmalloc(size, lflags | __GFP_HIGHMEM);
+	BUILD_BUG_ON(!ISP2(PAGE_SIZE));
 
-	/* Resulting allocated memory will be page aligned */
-	ASSERT(IS_P2ALIGNED(ptr, PAGE_SIZE));
+	gfp_t lflags = kmem_flags_convert(flags);
+	void *ptr = NULL;
+
+	const int pad_size = sizeof(kvalloc_endmarker_t) + (PAGE_SIZE - 1);
+	char* unaligned_ptr = spl_kvmalloc(size + pad_size, lflags);
+	if (likely(unaligned_ptr != NULL)) {
+		ptr = (void*)P2ROUNDUP_TYPED(unaligned_ptr, PAGE_SIZE, uintptr_t);
+		kvalloc_endmarker_t* mkr = (kvalloc_endmarker_t*)((char*)ptr + size);
+		mkr->magic = kvmagic_alive;
+		mkr->real_allocation_size = size;
+
+		/* Callers expect returned pointer to be page aligned */
+		ASSERT(IS_P2ALIGNED(ptr, PAGE_SIZE));
```

Comment:
Since the new kv_alloc() supports arbitrary and variable alignment - though it's currently hardcoded at PAGE_SIZE - it could take alignment as an parameter and then spl_slab_alloc() would pass the expected alignment for the objects inside the slab.  This would be a bit less wasteful than always allocating an extra ~ PAGE_SIZE.

I didn't want to introduce risk (like, are any slab creators lying/mistaken about their own alignment requirements?) but it'd be easy and uninvasive here.

Reply:
Wasting a whole extra page for every allocation seems really unfortunate.  (We do this in other places as well, which I'd also like to address.).  Could we do something like say that we support alignment of up to 64 bytes, and assert that no higher alignment is requested?

Reply:
No prob, I've been running with a version today which does minimal alignment for whatever object alignment is requested.  It's less tested than the static-alignment version but seems good.
I'll push it soon.